File: Reserve Your Spot/Services/BusinessService.cs
using Reserve_Your_Spot.Models;

namespace Reserve_Your_Spot.Services;

public class BusinessService : IBusinessService
{
    private readonly SupabaseService _supabase;

    public BusinessService(SupabaseService supabase)
    {
        _supabase = supabase;
    }

    public async Task<List<Business>> GetNearbyBusinessesAsync(double lat, double lng, double radiusKm = 10)
    {
        try
        {
            var all = await _supabase.GetListAsync<Business>("businesses", "select=*&is_approved=eq.true");
            foreach (var b in all)
                b.DistanceKm = CalculateDistance(lat, lng, b.Latitude, b.Longitude);
            return all.Where(b => b.DistanceKm <= radiusKm)
                      .OrderBy(b => b.DistanceKm)
                      .ToList();
        }
        catch { return new(); }
    }

    public async Task<List<Business>> GetFeaturedBusinessesAsync()
    {
        try
        {
            return await _supabase.GetListAsync<Business>(
                "businesses", "select=*&is_approved=eq.true&is_featured=eq.true&order=rating.desc&limit=10");
        }
        catch { return new(); }
    }

    public async Task<List<Business>> SearchBusinessesAsync(string query, string? categorySlug = null)
    {
        try
        {
            var filter = $"select=*&is_approved=eq.true&name=ilike.*{Uri.EscapeDataString(query)}*";
            var all = await _supabase.GetListAsync<Business>("businesses", filter);
            if (!string.IsNullOrEmpty(categorySlug))
                all = all.Where(b => b.Category?.Slug == categorySlug || b.CategoryId == categorySlug).ToList();
            return all;
        }
        catch { return new(); }
    }

    public async Task<Business?> GetBusinessByIdAsync(string id)
    {
        try { return await _supabase.GetSingleAsync<Business>("businesses", $"id=eq.{id}&select=*"); }
        catch { return null; }
    }

    public async Task<Business?> GetBusinessByOwnerAsync(string ownerId)
    {
        try { return await _supabase.GetSingleAsync<Business>("businesses", $"owner_id=eq.{ownerId}&select=*"); }
        catch { return null; }
    }

    public async Task<List<Service>> GetServicesAsync(string businessId)
    {
        try { return await _supabase.GetListAsync<Service>("services", $"business_id=eq.{businessId}&is_active=eq.true&select=*"); }
        catch { return new(); }
    }

    public async Task<List<Staff>> GetStaffAsync(string businessId)
    {
        try { return await _supabase.GetListAsync<Staff>("staff", $"business_id=eq.{businessId}&is_active=eq.true&select=*"); }
        catch { return new(); }
    }

    public async Task<List<Review>> GetReviewsAsync(string businessId)
    {
        try { return await _supabase.GetListAsync<Review>("reviews", $"business_id=eq.{businessId}&select=*&order=created_at.desc"); }
        catch { return new(); }
    }

    public async Task<bool> ToggleFavouriteAsync(string customerId, string businessId)
    {
        try
        {
            var isFav = await IsFavouriteAsync(customerId, businessId);
            if (isFav)
                await _supabase.DeleteAsync("favourites", $"customer_id=eq.{customerId}&business_id=eq.{businessId}");
            else
                await _supabase.InsertAsync<object>("favourites", new { customer_id = customerId, business_id = businessId });
            return !isFav;
        }
        catch { return false; }
    }

    public async Task<bool> IsFavouriteAsync(string customerId, string businessId)
    {
        try
        {
            var list = await _supabase.GetListAsync<object>(
                "favourites", $"customer_id=eq.{customerId}&business_id=eq.{businessId}&select=id");
            return list.Count > 0;
        }
        catch { return false; }
    }

    public async Task<List<Business>> GetFavouritesAsync(string customerId)
    {
        try
        {
            // Get favourite business IDs then load businesses
            var favs = await _supabase.GetListAsync<FavouriteRow>("favourites", $"customer_id=eq.{customerId}&select=business_id");
            var ids = favs.Select(f => f.BusinessId).ToList();
            var result = new List<Business>();
            foreach (var id in ids)
            {
                var b = await GetBusinessByIdAsync(id);
                if (b != null) result.Add(b);
            }
            return result;
        }
        catch { return new(); }
    }

    public async Task<Business?> CreateBusinessAsync(Business business)
    {
        try { return await _supabase.InsertAsync<Business>("businesses", business); }
        catch { return null; }
    }

    public async Task UpdateBusinessAsync(Business business)
    {
        try { await _supabase.UpdateAsync("businesses", $"id=eq.{business.Id}", business); }
        catch { /* swallow */ }
    }

    public async Task<Service?> AddServiceAsync(Service service)
    {
        try { return await _supabase.InsertAsync<Service>("services", service); }
        catch { return null; }
    }

    public async Task UpdateServiceAsync(Service service)
    {
        try { await _supabase.UpdateAsync("services", $"id=eq.{service.Id}", service); }
        catch { }
    }

    public async Task DeleteServiceAsync(string serviceId)
    {
        try { await _supabase.DeleteAsync("services", $"id=eq.{serviceId}"); }
        catch { }
    }

    public async Task<Staff?> AddStaffAsync(Staff staff)
    {
        try { return await _supabase.InsertAsync<Staff>("staff", staff); }
        catch { return null; }
    }

    public async Task UpdateStaffAsync(Staff staff)
    {
        try { await _supabase.UpdateAsync("staff", $"id=eq.{staff.Id}", staff); }
        catch { }
    }

    public async Task DeleteStaffAsync(string staffId)
    {
        try { await _supabase.DeleteAsync("staff", $"id=eq.{staffId}"); }
        catch { }
    }

    private static double CalculateDistance(double lat1, double lon1, double lat2, double lon2)
    {
        const double R = 6371;
        var dLat = ToRad(lat2 - lat1);
        var dLon = ToRad(lon2 - lon1);
        var a = Math.Sin(dLat / 2) * Math.Sin(dLat / 2)
              + Math.Cos(ToRad(lat1)) * Math.Cos(ToRad(lat2))
              * Math.Sin(dLon / 2) * Math.Sin(dLon / 2);
        return R * 2 * Math.Atan2(Math.Sqrt(a), Math.Sqrt(1 - a));
    }

    private static double ToRad(double deg) => deg * Math.PI / 180;

    private class FavouriteRow { public string BusinessId { get; set; } = string.Empty; }
}

File: booklocal-mobile/services/businessService.ts
import { supabase } from './supabase';
import { Business, Service, StaffMember, Review } from '../types';
import { DEFAULT_SEARCH_RADIUS_KM } from '../constants/AppConstants';

function haversineKm(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const R = 6371;
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLon = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLon / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

export const businessService = {
  async getNearbyBusinesses(
    lat: number,
    lng: number,
    radiusKm = DEFAULT_SEARCH_RADIUS_KM,
    categorySlug?: string
  ): Promise<Business[]> {
    let query = supabase
      .from('businesses')
      .select('*, category:categories(*)')
      .eq('is_approved', true);

    if (categorySlug && categorySlug !== 'all') {
      const { data: cat } = await supabase
        .from('categories')
        .select('id')
        .eq('slug', categorySlug)
        .single();
      if (cat) query = query.eq('category_id', cat.id);
    }

    const { data, error } = await query;
    if (error || !data) return [];

    return (data as Business[])
      .map((b) => ({
        ...b,
        distance_km: haversineKm(lat, lng, b.latitude, b.longitude),
      }))
      .filter((b) => b.distance_km! <= radiusKm)
      .sort((a, b) => a.distance_km! - b.distance_km!);
  },

  async searchBusinesses(
    query: string,
    categorySlug?: string,
    lat?: number,
    lng?: number
  ): Promise<Business[]> {
    let dbQuery = supabase
      .from('businesses')
      .select('*, category:categories(*)')
      .eq('is_approved', true)
      .ilike('name', `%${query}%`);

    if (categorySlug && categorySlug !== 'all') {
      const { data: cat } = await supabase
        .from('categories')
        .select('id')
        .eq('slug', categorySlug)
        .single();
      if (cat) dbQuery = dbQuery.eq('category_id', cat.id);
    }

    const { data, error } = await dbQuery.limit(30);
    if (error || !data) return [];

    const results = data as Business[];
    if (lat != null && lng != null) {
      return results
        .map((b) => ({ ...b, distance_km: haversineKm(lat, lng, b.latitude, b.longitude) }))
        .sort((a, b) => a.distance_km! - b.distance_km!);
    }
    return results;
  },

  async getBusinessBySlug(slug: string): Promise<Business | null> {
    const { data, error } = await supabase
      .from('businesses')
      .select('*, category:categories(*), services(*), staff(*), reviews(*, customer:users(*))')
      .eq('slug', slug)
      .single();
    if (error) return null;
    return data as Business;
  },

  async getBusinessById(id: string): Promise<Business | null> {
    const { data, error } = await supabase
      .from('businesses')
      .select('*, category:categories(*), services(*), staff(*), reviews(*, customer:users(*))')
      .eq('id', id)
      .single();
    if (error) return null;
    return data as Business;
  },

  async getBusinessByOwnerId(ownerId: string): Promise<Business | null> {
    const { data, error } = await supabase
      .from('businesses')
      .select('*')
      .eq('owner_id', ownerId)
      .single();
    if (error) return null;
    return data as Business;
  },

  async createBusiness(business: Partial<Business>): Promise<{ data: Business | null; error: string | null }> {
    const { data, error } = await supabase
      .from('businesses')
      .insert(business)
      .select()
      .single();
    return { data: data as Business | null, error: error?.message ?? null };
  },

  async updateBusiness(id: string, updates: Partial<Business>): Promise<{ error: string | null }> {
    const { error } = await supabase.from('businesses').update(updates).eq('id', id);
    return { error: error?.message ?? null };
  },

  async getFeaturedBusinesses(): Promise<Business[]> {
    const { data } = await supabase
      .from('businesses')
      .select('*, category:categories(*)')
      .eq('is_approved', true)
      .eq('is_featured', true)
      .order('rating', { ascending: false })
      .limit(10);
    return (data as Business[]) ?? [];
  },

  async getServices(businessId: string): Promise<Service[]> {
    const { data } = await supabase
      .from('services')
      .select('*')
      .eq('business_id', businessId)
      .eq('is_active', true)
      .order('name');
    return (data as Service[]) ?? [];
  },

  async createService(service: Partial<Service>): Promise<{ error: string | null }> {
    const { error } = await supabase.from('services').insert(service);
    return { error: error?.message ?? null };
  },

  async updateService(id: string, updates: Partial<Service>): Promise<{ error: string | null }> {
    const { error } = await supabase.from('services').update(updates).eq('id', id);
    return { error: error?.message ?? null };
  },

  async deleteService(id: string): Promise<{ error: string | null }> {
    const { error } = await supabase.from('services').delete().eq('id', id);
    return { error: error?.message ?? null };
  },

  async getStaff(businessId: string): Promise<StaffMember[]> {
    const { data } = await supabase
      .from('staff')
      .select('*')
      .eq('business_id', businessId)
      .eq('is_active', true);
    return (data as StaffMember[]) ?? [];
  },

  async createStaff(member: Partial<StaffMember>): Promise<{ error: string | null }> {
    const { error } = await supabase.from('staff').insert(member);
    return { error: error?.message ?? null };
  },

  async updateStaff(id: string, updates: Partial<StaffMember>): Promise<{ error: string | null }> {
    const { error } = await supabase.from('staff').update(updates).eq('id', id);
    return { error: error?.message ?? null };
  },

  async getReviews(businessId: string): Promise<Review[]> {
    const { data } = await supabase
      .from('reviews')
      .select('*, customer:users(full_name, avatar_url)')
      .eq('business_id', businessId)
      .order('created_at', { ascending: false });
    return (data as Review[]) ?? [];
  },

  async createReview(review: Partial<Review>): Promise<{ error: string | null }> {
    const { error } = await supabase.from('reviews').insert(review);
    return { error: error?.message ?? null };
  },

  async toggleFavourite(customerId: string, businessId: string): Promise<boolean> {
    const { data: existing } = await supabase
      .from('favourites')
      .select('id')
      .eq('customer_id', customerId)
      .eq('business_id', businessId)
      .single();

    if (existing) {
      await supabase.from('favourites').delete().eq('id', existing.id);
      return false;
    } else {
      await supabase.from('favourites').insert({ customer_id: customerId, business_id: businessId });
      return true;
    }
  },

  async getFavourites(customerId: string): Promise<Business[]> {
    const { data } = await supabase
      .from('favourites')
      .select('business:businesses(*, category:categories(*))')
      .eq('customer_id', customerId);
    return (data?.map((f: any) => f.business) ?? []) as Business[];
  },

  async isFavourite(customerId: string, businessId: string): Promise<boolean> {
    const { data } = await supabase
      .from('favourites')
      .select('id')
      .eq('customer_id', customerId)
      .eq('business_id', businessId)
      .single();
    return !!data;
  },

  async getLoyaltyCard(customerId: string, businessId: string) {
    const { data } = await supabase
      .from('loyalty_cards')
      .select('*, business:businesses(name, logo_url)')
      .eq('customer_id', customerId)
      .eq('business_id', businessId)
      .single();
    return data;
  },

  async getCustomerLoyaltyCards(customerId: string) {
    const { data } = await supabase
      .from('loyalty_cards')
      .select('*, business:businesses(name, logo_url, brand_color)')
      .eq('customer_id', customerId);
    return data ?? [];
  },

  generateSlug(name: string): string {
    return name
      .toLowerCase()
      .replace(/[^a-z0-9\s-]/g, '')
      .replace(/\s+/g, '-')
      .replace(/-+/g, '-')
      .trim();
  },
};
